File: Algorithms/cpp/005 - Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s0) {
        int size = s0.size();
        int n = 2 * size +2;
        int *p = new int[n];
        string s = "@#";
        for(int i=0; i < size; ++i){
            s += s0[i];
            s += '#';
        }
        int mx = 0, id = 0;
        int ans_value = 0, ans_i = 0;
        for(int i=0; i<n; i++){
            p[i] = mx > i ? min(p[2*id-i], mx-i) : 1;
            while(i - p[i] >=0 && i + p[i] < n && s[i+p[i]] == s[i-p[i]]){
                p[i]++;
            }
            if (p[i] + i -1 > mx) {
                mx = p[i] + i - 1;
                id = i;
            }
            if (p[i] > ans_value){
                ans_value = p[i];
                ans_i = i;
            }
        }


        int start = (ans_i - ans_value)/2;
        return s0.substr(start, ans_value-1);
    }
};

File: Algorithms/python/005 - Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        return self.manacher(s)
    

    def manacher(self, s0 : str) -> list:
        s = '@#' + '#'.join(s0) + '#'
        l = len(s)
        r = [0] * l
        mx, pos = 0, 0
        ans_i = 0
        ans_value = 0
        for i in range(l):
            if i > mx:
                r[i] = 1
            else:
                r[i] = min(r[2 * pos - i], mx - i)
        
            while i - r[i] >= 0 and i + r[i] < l and s[i+r[i]] == s[i-r[i]]:
                r[i] += 1
            
            if r[i] + i - 1 > mx:
                mx = r[i] + i - 1
                pos = i
            
            if r[i] > ans_value:
                ans_i = i
                ans_value = r[i]
            
        start = int((ans_i - ans_value)/2)
        end = int(start + ans_value - 1) 
        return s0[start:end]
